design/bmc-address.md

```diff
@@ -0,0 +1,126 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# bmc-address
+
+## Status
+
+implementable
+
+## Table of Contents
+
+<!--ts-->
+   * [bmc-address](#bmc-address)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Alternatives](#alternatives)
+      * [References](#references)
+
+<!-- Added by: dhellmann, at: 2019-03-04T09:33-0500 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the way users will provide the network location
+of the management controller on a host.
+
+## Motivation
+
+We need to document how we are going to specify the location of a BMC,
+including how to tell its type, IP address, port number, and
+potentially the path.
+
+### Goals
+
+1. To agree on an address specification system, including assumptions
+   that can be made when parts of an address are left out.
+
+### Non-Goals
+
+1. To list the scheme to use for every type of controller.
+2. To specify the user interface for entering address information.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+For each BMC, we need to know the type, IP, port, and optionally the
+path to use to communicate with the controller.
+
+We can collect all of this information using a single "address" field,
+if we allow that field to contain partial or full URLs.
+
+For each type of controller, we can assume we know the URL scheme used
+to communicate with it (HTTP, HTTPS, etc.). Therefore the scheme in a
+URL provided by the user is redundant, and we can use that portion of
+the URL to specify the controller type. For example:
+
+    ipmi://192.168.111.1
+
+Initially, we will only support IPMI controllers, so we do not need
+users to specify the type.  If the field only contains a network
+address, we can assume that the controller uses IPMI on the standard
+port, 623. Therefore this would be equivalent to the previous example:
+
+    192.168.111.1
```

Comment:
This assumption makes sense to me, and requiring the user to specify a `type` in the future when more controllers are supported (either via the same field or a dropdown in the UI) seems natural.

Do BMCs typically expose multiple controller types at once? Are certain types better than others? I assume so, but if not, would we be able to make our backend logic even smarter and automatically check across multiple controller types to avoid asking the user to specify the `type` in addition to the IP? Are there risks to doing that, like the user getting blocked by the BMC?

Reply:
It seems likely that a given controller would support a standard and vendor-specific interfaces. I think for our purposes, though, we would pick one and tell the user to always select that one based on the test matrix and what we find most reliable. I don't know if we have any way of detecting the type, since we need the user to tell us where it is and that location may be more complex than just an IP and port (see the redfish path discussion in other comments), but perhaps one of the folks more familiar with Ironic can give us more info.

Reply:
Yes, one controller can support several protocols (usually on different ports). Also worth knowing that one controller can support several nodes (hence the problem with Redfish).

Reply:
> select that one based on the test matrix and what we find most reliable

May be the other criterion could be based on the feature set required.

>  if we have any way of detecting the type

We could perhaps do a port scan. For IPMI I recall something like that being in TripleO.

Also, I think the MaaS project has quite developed BMC discovery capabilities.

Reply:
I don't think we want the user having to choose which driver to use for a given host. We're going to have a limited set of hardware to support, so we should be able to tell them the best setting to use in each case.

Reply:
Discovery of hosts is something we've put off for future releases, but if we can identify the right BMC that would be a good addition at that point.